marketing.js

```diff
@@ -0,0 +1,229 @@
+'use strict'
+
+const AutoDetectDecoderStream = require('autodetect-decoder-stream')
+const crypto = require('crypto')
+const csv = require('csv-parser')
+const fs = require('fs')
+const Promise = require('bluebird')
+const request = require('request-promise')
+const s3 = require('s3')
+
+fs.unlinkAsync = Promise.promisify(fs.unlink)
+
+const TIMESTAMP = /^(20[1-9][0-9])-([01][0-9])-([0-3][1-9])-([0-2][0-9])-([0-5][0-9])$/
+const EVENT = /^mktg-([a-z]+_[a-z]+)$/
+
+const AWS_ACCESS_KEY = process.env.FXA_AWS_ACCESS_KEY
+const AWS_SECRET_KEY = process.env.FXA_AWS_SECRET_KEY
+const AWS_S3_BUCKET = '*** TODO ***'
+const AWS_S3_PREFIX = '*** TODO ***'
+
+const MAX_EVENTS_PER_BATCH = 10
+const HMAC_KEY = process.env.FXA_AMPLITUDE_HMAC_KEY
+const API_KEY = process.env.FXA_AMPLITUDE_API_KEY
+
+if (process.argv.length !== 3) {
+  console.error(`Usage: ${process.argv[1]} {YYYY-MM-DD-hh-mm | LOCAL PATH}`)
+  console.error('If specifying YYYY-MM-DD-hh-mm as the arg, note that the script will try to send events')
+  console.error('for all times from YYYY-MM-DD-hh-mm to the most recent available in S3. If any times in')
+  console.error('that range are missing, they will be skipped without failing the process.')
+  process.exit(1)
+}
+
+if (! HMAC_KEY || ! API_KEY) {
+  console.error('Error: You must set FXA_AMPLITUDE_HMAC_KEY and FXA_AMPLITUDE_API_KEY environment variables')
+  process.exit(1)
+}
+
+processData()
+  .then(count => {
+    console.log(`Sent ${count} events!`)
+    process.exit(0)
+  })
+  .catch(error => {
+    console.error(error.stack)
+    process.exit(1)
+  })
+
+function processData () {
+  const timeParts = TIMESTAMP.exec(process.argv[2])
+
+  if (timeParts && timeParts.length === 6) {
+    if (! AWS_ACCESS_KEY || ! AWS_SECRET_KEY) {
+      console.error('Error: You must set AWS_ACCESS_KEY and AWS_SECRET_KEY environment variables')
+      process.exit(1)
+    }
+
+    return processDataFromS3(timeParts.slice(1))
+      .then(counts => counts.reduce((sum, count) => sum + count, 0))
+  }
+
+  return processStream(fs.createReadStream(process.argv[2]))
+}
+
+function processStream (stream) {
+  let eventCount = 0, batch = [], error
+
+  return new Promise((resolve, reject) => {
+    stream
+      .pipe(new AutoDetectDecoderStream())
```

Comment:
The example CSV we were sent was UTF-16. This step detects the encoding using the BOM if present (it was in the sample data) and transforms to JS strings.